sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala

```diff
@@ -232,12 +232,149 @@ final class Decimal extends Ordered[Decimal] with Serializable {
     }
   }
 
+  /**
+   * @return the Long value that is equal to the rounded decimal.
+   * @throws ArithmeticException if checkOverflow is true and
+   *                             the decimal too big to fit in Long type.
+   */
+  def toLong(checkOverflow: Boolean): Long = {
```

Comment:
Then all the other function calls `toLong` should be at least add parentheses as `toLong()`. Since Scala only allows Arity-0 https://docs.scala-lang.org/style/method-invocation.html#arity-0 to omit parentheses.
My two concerns here:
1. The existing external code calling `Decimal.toLong` will fail
2. The usage will be different from the trait `Numeric`

Reply:
BTW, renaming to `toLongExact` won't accurate either. For example, `toLongExact(1.1)` should be false, while we are actually doing rounding in toLong.

Reply:
for 2, anyway, the usage is already different and here we're not in a `Numeric`-like class. On 1, I am not sure it is a problem. `Decimal` is `Unstable` and this patch will go in Spark 3.0, so it is a major release (best place where to put a breaking change!). And the benefit to avoid extra function calls and a lot of duplicated code is worth the change IMHO.

cc @cloud-fan @maropu for their opinion on this too, but I feel quite strong about this.

Reply:
Actually, I am not super comfortable with the code changes in Decimal.scala here.
I did this to address comments in https://github.com/apache/spark/pull/25461#discussion_r315177196 . 
How about just adding a new method `roundToLong`. The name is same as
https://github.com/google/guava/blob/master/guava/src/com/google/common/math/DoubleMath.java#L156 . So that we can leave toLong as it is.

Reply:
We can't break public classes just to make it easier to write code in Spark.

Reply:
I think we can remove the additional function call by codegen. We can remove `def toLong(checkOverflow: Boolean): Long`, and in the codegen:
```
if (nullOnOverFlow) decimal.toLong else decimal.roundToLong
```

Reply:
+1 for @cloud-fan 's suggestion. At least we remove the extra method call.